GliaWidgets/SecureConversations/SecureConversations.FileUploadList.swift

```diff
@@ -0,0 +1,965 @@
+        for element in elements {
+            self.append(item: element, identified: element.id)
+        }
+    }
+}
+
+extension IdCollection where Item: Identifiying, Item.ID == Identifier {
+    init(_ elements: [Item]) {
+        for element in elements {
+            self.append(item: element, identified: element.id)
+        }
+    }
+}
+
+extension SecureConversations {
+    class FileUploadListViewModel: ViewModel {
+        var action: ((Action) -> Void)?
+        var delegate: ((DelegateEvent) -> Void)?
+
+        enum Event {}
+        enum Action {}
+
+        enum DelegateEvent {
+            case renderProps(SecureConversations.FileUploadListView.Props)
+        }
+
+        let environment: Environment
+
+        init(environment: Environment) {
+            self.environment = environment
+            environment.uploader.state.addObserver(self) { [weak self] state, _ in
+                self?.onUploaderStateChanged(state)
+            }
+            environment.uploader.limitReached.addObserver(self) { [weak self] limitReached, _ in
+                // TODO: MOB-1722
+            }
+        }
+
+        func event(_ event: Event) {}
+
+        func reportChange() {
+            delegate?(.renderProps(props()))
+        }
+
+        func observeUpload(
+            _ upload: FileUpload,
+            instance: FileUploadListViewModel
+        ) {
+            // We need to observe state of the FileUpload,
+            // and trigger Props generation.
+            var progressObservable: ObservableValue<Double>?
+
+            func handleState(
+                _ state: FileUpload.State,
+                instance: FileUploadListViewModel
+            ) {
+                progressObservable?.removeObserver(instance)
+                switch state {
+                case .none, .uploaded, .error:
+                    break
+                case .uploading(progress: let progress):
+                    progressObservable = progress
+                    progressObservable?.addObserver(instance) { [weak instance] _, _ in
+                        instance?.reportChange()
+                    }
+                }
+
+                instance.reportChange()
+            }
+
+            upload.state.addObserver(instance) { [weak instance] state, _ in
+                guard let instance = instance else { return }
+                handleState(state, instance: instance)
+            }
+
+            // There is a case, where state is already set to `.uploading` case,
+            // that's why we evaluate state immediatelty.
+            handleState(upload.state.value, instance: instance)
+        }
+
+        func onUploaderStateChanged(_ state: FileUploader.State) {
+            reportChange()
+        }
+
+        func removeUpload(_ upload: FileUpload) {
+            upload.state.removeObserver(self)
+            environment.uploader.removeUpload(upload)
+            reportChange()
+//            action?(.removeUpload(upload))
```

Comment:
Will it be uncommented or can be deleted?

Reply:
Removed 3ecc2f54eca61a1d236e51e3048b292b4ef28583.